content/issue-2/write-a-sql-optimizer-using-egg.md

```diff
@@ -0,0 +1,326 @@
+4. Plan Node: similar to an operator, but it operates on a table instead of a value. A SQL statement is transformed into a query plan tree composed of several nodes. Each plan node contains not only child nodes (`child`, `left`, `right`) which are plans, but also several parameters composed of expressions.
+
+Let's say we have a SQL statement like this:
+
+```sql
+SELECT users.name, count(commits.id)
+FROM users, repos, commits
+WHERE users.id = commits.user_id
+	AND repos.id = commits.repo_id
+	AND repos.name = 'RisingLight'
+	AND commits.time BETWEEN date '2022-01-01' AND date '2022-12-31'
+GROUP BY users.name
+HAVING count(commits.id) >= 10
+ORDER BY count(commits.id) DESC
+LIMIT 10
+```
+
+Its query plan can be represented in our language as follows:
+
+```sql
+(proj (list $1.2 (count $3.1))
+  (limit 10 0
+    (order (list (desc (count $3.1)))
+      (filter (>= (count $3.1) 10)
+        (agg (list (count $3.1)) (list $1.2)
+          (filter (and (and (and (= $1.1 $3.2) (= $2.1 $3.3)) (= $2.2 'RisingLight'))
+                    (and (>= $3.4 '2022-01-01') (<= $3.4 '2022-12-31')))
+            (join inner true
+              (join inner true
+                (scan $1 (list $1.1 $1.2 $1.3))
+                (scan $2 (list $2.1 $2.2 $2.3)))
+              (scan $3 (list $3.1 $3.2 $3.3 $3.4 $3.5))
+)))))))
+```
+
+The next step is to define a series of rules to perform equivalent transformations on the expression above, and eventually find the representation with the smallest cost as the optimized result.
+
+# Define Rewrite Rules
+
+In this section, we will use the API provided by Egg to describe optimization rules. Due to space limitations, we will choose several representative ones for a brief introduction. They are:
+
+- Expression simplification
+- Constant folding
+- Predicate pushdown
+- HashJoin
+- Join reordering
+- Cost function
+
+## Expression Simplification
+
+Firstly, we simplify expressions in the plan nodes.
+
+For example, the multiplication-to-shift rule mentioned at the beginning can be described in Egg using the [`rewrite`](https://docs.rs/egg/0.9.3/egg/macro.rewrite.html) macro as follows:
+
+```rust
+rewrite!("mul2-to-shl1"; "(* ?a 2)" => "(<< ?a 1)")
+```
+
+The strings on either side of `=>` represent the expression patterns before and after the transformation, where `?a` represents a variable that can be replaced with any expression. When the runner finds an expression that matches this pattern, it generates a new expression according to the pattern on the right and unions it with the original node to form an equivalence class. By the way, `=>` can also be replaced with `<=>` to indicate a bidirectional transformation.
+
+For some rules, they require certain conditions to be met before execution. For example, the following rule is only valid when the divisor is not zero:
```

Comment:
Nit-Pick suggestion -- but probably, I think given the language used so far, it seems you are talking to me ... and I am enjoying the conversation :)
```suggestion
Some rules require certain conditions to be met before execution. For example, the following rule is only valid when the divisor is not zero:
```

Reply:
Thanks!